onmt/utils/loss.py

```diff
@@ -184,22 +191,59 @@ def __call__(self,
-        shard_state = self._make_shard_state(batch, output, trunc_range, attns)
-        if shard_size == 0:
-            loss, stats = self._compute_loss(batch, **shard_state)
-            return loss / float(normalization), stats
-        batch_stats = onmt.utils.Statistics()
-        for shard in shards(shard_state, shard_size):
-            loss, stats = self._compute_loss(batch, **shard)
-            loss.div(float(normalization)).backward()
-            batch_stats.update(stats)
-        return None, batch_stats
+        # take into account here the tgt_shift_index (0 / 1 = LM/NMT)
+        trunc_range = (trunc_start + self.tgt_shift_index,
+                       trunc_start + trunc_size)
+        target = batch['tgt'][trunc_range[0]:trunc_range[1],
+                              :, 0].view(-1)
+
+        if self.copy_attn:
+            align = batch['alignment'][trunc_range[0]:trunc_range[1]].view(-1)
+            loss, stats = self._compute_copy_loss(batch, output, target,
+                                                  align, attns)
+        else:
+
+            scores = self.generator(self._bottle(output))
+            loss = self.criterion(scores, target)
+
+            if self.lambda_align != 0.0:
+                align_head = attns['align']
+                if align_head.dtype != loss.dtype:  # Fix FP16
+                    align_head = align_head.to(loss.dtype)
+                align_idx = batch['align']
+                pad_tgt_size, batch_size, _ = batch['tgt'].size()
+                pad_src_size, _, _ = batch['src'].size()
+                align_matrix_size = [batch_size, pad_tgt_size, pad_src_size]
+                ref_align = onmt.utils.make_batch_align_matrix(
+                    align_idx, align_matrix_size, normalize=True
+                )
+                ref_align = ref_align[:, trunc_range[0]:trunc_range[1], :]
+                if ref_align.dtype != loss.dtype:
+                    ref_align = ref_align.to(loss.dtype)
+                align_loss = self._compute_alignement_loss(
+                    align_head=align_head, ref_align=ref_align)
+                loss += align_loss
+
+            stats = self._stats(loss.item(), scores, target)
+
+        if self.lambda_coverage != 0.0:
+            coverage_loss = self._compute_coverage_loss(
+                std_attn=attns['std'], coverage_attn=attns['coverage'])
+            loss += coverage_loss
+
+        if self.normalization == "tokens":
+            normfactor = batch['tgt'][:,
+                                      :,
+                                      0].ne(self.padding_idx).sum()
+        elif self.normalization == "sents":
+            normfactor = batch['tgt'].size(1)
+
+        return loss / float(normfactor), stats
 
     def _stats(self, loss, scores, target):
         """
         Args:
-            loss (:obj:`FloatTensor`): the loss computed by the loss criterion.
+            loss (int): the loss computed by the loss criterion.
```

Comment:
int?

Reply:
yes .sum().item() is done before passing now to avoid .clone()